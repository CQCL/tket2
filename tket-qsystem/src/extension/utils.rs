//! This module defines the "tket.qsystem.utils" extension, which includes the
//! utility functions available for Quantinuum systems.

use std::sync::{Arc, Weak};

use derive_more::derive::Display;
use hugr::{
    builder::{BuildError, Dataflow},
    extension::{
        prelude::UnwrapBuilder,
        simple_op::{try_from_name, MakeOpDef, MakeRegisteredOp},
        ExtensionId, ExtensionRegistry, OpDef, SignatureFunc, Version, PRELUDE,
    },
    std_extensions::arithmetic::int_types::int_type,
    type_row,
    types::Signature,
    Extension, Wire,
};
use lazy_static::lazy_static;
use strum::{EnumIter, EnumString, IntoStaticStr};

/// The extension ID for the utils extension.
pub const EXTENSION_ID: ExtensionId = ExtensionId::new_unchecked("tket.qsystem.utils");
/// The version of the "tket.qsystem.utils" extension.
pub const EXTENSION_VERSION: Version = Version::new(0, 3, 0);

lazy_static! {
    /// The "tket.qsystem.utils" extension.
    pub static ref EXTENSION: Arc<Extension> = {
         Extension::new_arc(EXTENSION_ID, EXTENSION_VERSION, |ext, ext_ref| {
            UtilsOp::load_all_ops( ext, ext_ref).unwrap();
        })
    };

    /// Extension registry including the "tket.qsystem.utils" extension and
    /// dependencies.
    pub static ref REGISTRY: ExtensionRegistry = ExtensionRegistry::new([
        EXTENSION.to_owned(),
        PRELUDE.to_owned(),
    ]);
}

#[derive(
    Clone,
    Copy,
    Debug,
    serde::Serialize,
    serde::Deserialize,
    Hash,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    EnumIter,
    IntoStaticStr,
    EnumString,
    Display,
)]
#[non_exhaustive]
/// The operations provided by the utils extension.
pub enum UtilsOp {
    /// `fn get_current_shot() -> usize`
    GetCurrentShot,
}

impl MakeOpDef for UtilsOp {
    fn opdef_id(&self) -> hugr::ops::OpName {
        <&'static str>::from(self).into()
    }

    fn init_signature(&self, _extension_ref: &std::sync::Weak<Extension>) -> SignatureFunc {
        match self {
            UtilsOp::GetCurrentShot => Signature::new(type_row![], int_type(6)),
        }
        .into()
    }

    fn from_def(op_def: &OpDef) -> Result<Self, hugr::extension::simple_op::OpLoadError> {
        try_from_name(op_def.name(), op_def.extension_id())
    }

    fn extension(&self) -> ExtensionId {
        EXTENSION_ID
    }

    fn extension_ref(&self) -> std::sync::Weak<Extension> {
        Arc::downgrade(&EXTENSION)
    }

    fn description(&self) -> String {
        match self {
            UtilsOp::GetCurrentShot => "Get current shot number.",
        }
        .to_string()
    }
}

impl MakeRegisteredOp for UtilsOp {
    fn extension_id(&self) -> ExtensionId {
        EXTENSION_ID
    }

    fn extension_ref(&self) -> Weak<Extension> {
        Arc::downgrade(&EXTENSION)
    }
}

/// An extension trait for [Dataflow] providing methods to add
/// "tket.qsystem.utils" operations.
pub trait UtilsOpBuilder: Dataflow + UnwrapBuilder {
    /// Add a "tket.qsystem.utils.GetCurrentShot" op.
    fn add_get_current_shot(&mut self) -> Result<Wire, BuildError> {
        Ok(self
            .add_dataflow_op(UtilsOp::GetCurrentShot, [])?
            .out_wire(0))
    }
}

impl<D: Dataflow> UtilsOpBuilder for D {}

#[cfg(test)]
mod test {
    use std::sync::Arc;

    use hugr::HugrView;
    use hugr::{
        builder::{DataflowHugr, FunctionBuilder},
        extension::simple_op::MakeExtensionOp,
    };
    use strum::IntoEnumIterator;

    use super::*;

    fn get_opdef(op: UtilsOp) -> Option<&'static Arc<OpDef>> {
        EXTENSION.get_op(&op.op_id())
    }

    #[test]
    fn create_extension() {
        assert_eq!(EXTENSION.name(), &EXTENSION_ID);

        for o in UtilsOp::iter() {
            assert_eq!(UtilsOp::from_def(get_opdef(o).unwrap()), Ok(o));
        }
    }

    #[test]
    fn get_current_shot() {
        let hugr = {
            let mut func_builder = FunctionBuilder::new(
                "get_current_shot",
                Signature::new(vec![], vec![int_type(6)]),
            )
            .unwrap();
            let shot = func_builder.add_get_current_shot().unwrap();
            func_builder.finish_hugr_with_outputs([shot]).unwrap()
        };
        hugr.validate().unwrap()
    }
}
