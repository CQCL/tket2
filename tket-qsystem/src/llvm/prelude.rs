//! LLVM lowering implementations for HUGR prelude operations.

use std::hash::{DefaultHasher, Hash as _, Hasher as _};
use tket::hugr::{self, llvm::inkwell};

use anyhow::{anyhow, ensure, Result};
use hugr::llvm::emit::func::EmitFuncContext;
use hugr::llvm::extension::prelude::PreludeCodegen;
use hugr::llvm::types::TypingSession;
use inkwell::attributes::{Attribute, AttributeLoc};
use inkwell::module::Linkage;
use inkwell::types::{BasicType, IntType};
use inkwell::values::{BasicValue, BasicValueEnum, FunctionValue, StructValue};
use inkwell::AddressSpace;
use tket::hugr::{HugrView, Node};

#[derive(Clone)]
/// Codegen extension for the QIS prelude.
pub struct QISPreludeCodegen;

impl PreludeCodegen for QISPreludeCodegen {
    fn usize_type<'c>(&self, session: &TypingSession<'c, '_>) -> IntType<'c> {
        session.iw_context().i64_type()
    }

    /// Return the llvm type of [`tket::hugr::extension::prelude::qb_t`].
    fn qubit_type<'c>(&self, session: &TypingSession<'c, '_>) -> impl BasicType<'c> {
        session.iw_context().i64_type()
    }

    /// A panic ends all shots, this is done by using a return code > 1000.
    /// The lowering of the panic HUGR op shifts the return code held by the op by 1000.
    /// So valid values for `signal` in the panic HUGR op are 1-(2^31 - 1001).
    fn emit_panic<H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<H>,
        err: BasicValueEnum,
    ) -> Result<()> {
        let err = StructValue::try_from(err)
            .map_err(|()| anyhow!("emit_panic: Expected err to have struct type"))?;
        let return_code = ctx.builder().build_extract_value(err, 0, "")?;
        // TODO check if code is in valid range 1-(2^31 - 1001), otherwise error.
        // https://github.com/quantinuum-dev/eldarion/issues/287
        let return_code = {
            let shift = ctx.iw_context().i32_type().const_int(1000, false);
            ctx.builder()
                .build_int_add(return_code.into_int_value(), shift, "shift_code")?
        };

        let msg = ctx.builder().build_extract_value(err, 1, "")?;
        let panic = Self::get_panic(ctx)?;
        ctx.builder()
            .build_call(panic, &[return_code.into(), msg.into()], "")?;
        Ok(())
    }

    /// An exit ends the current shot using a return code `rc`, where `1 <= rc <= 1000`.
    fn emit_exit<H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<H>,
        err: BasicValueEnum,
    ) -> Result<()> {
        let err = StructValue::try_from(err)
            .map_err(|()| anyhow!("emit_exit: Expected err to have struct type"))?;
        let return_code = ctx.builder().build_extract_value(err, 0, "")?;
        // TODO check if code is in valid range 1-1000, otherwise error.
        // https://github.com/quantinuum-dev/eldarion/issues/287
        let msg = ctx.builder().build_extract_value(err, 1, "")?;
        let panic = Self::get_panic(ctx)?;
        ctx.builder()
            .build_call(panic, &[return_code.into(), msg.into()], "")?;
        Ok(())
    }

    fn emit_print<H: HugrView<Node = Node>>(
        &self,
        _ctx: &mut EmitFuncContext<H>,
        _text: inkwell::values::BasicValueEnum,
    ) -> Result<()> {
        Ok(())
    }

    fn emit_const_error<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        err: &hugr::extension::prelude::ConstError,
    ) -> Result<BasicValueEnum<'c>> {
        let err_ty = ctx
            .llvm_type(&hugr::extension::prelude::error_type())?
            .into_struct_type();
        let signal = err_ty
            .get_field_type_at_index(0)
            .unwrap()
            .into_int_type()
            .const_int(u64::from(err.signal), false);

        let message = emit_global_string(ctx, &err.message, "e_", "EXIT:INT:")?;
        let err = err_ty.const_named_struct(&[signal.into(), message]);
        Ok(err.into())
    }

    fn emit_const_string<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        str: &hugr::extension::prelude::ConstString,
    ) -> Result<BasicValueEnum<'c>> {
        emit_global_string(ctx, str.value(), "s_", "")
    }
}

impl QISPreludeCodegen {
    fn get_panic<'c, H: HugrView<Node = Node>>(
        ctx: &EmitFuncContext<'c, '_, H>,
    ) -> Result<FunctionValue<'c>> {
        let iwc = ctx.iw_context();
        let func_type = iwc.void_type().fn_type(
            &[
                iwc.i32_type().into(),
                iwc.i8_type().ptr_type(AddressSpace::default()).into(),
            ],
            false,
        );
        let f = ctx.get_extern_func("panic", func_type)?;
        let noreturn = Attribute::get_named_enum_kind_id("noreturn");
        debug_assert!(noreturn != 0);
        f.add_attribute(
            AttributeLoc::Function,
            iwc.create_enum_attribute(noreturn, 0),
        );

        Ok(f)
    }
}

/// Emit a global string constant with a unique name.
///
/// The symbol name is constructed with `symbol_prefix` and ``str`.
///
/// The string is prefixed with `str_prefix`.
pub fn emit_global_string<'c, H: HugrView<Node = Node>>(
    ctx: &EmitFuncContext<'c, '_, H>,
    str: impl AsRef<str>,
    symbol_prefix: impl AsRef<str>,
    str_prefix: impl AsRef<str>,
) -> Result<BasicValueEnum<'c>> {
    let (str, symbol_prefix, str_prefix) =
        (str.as_ref(), symbol_prefix.as_ref(), str_prefix.as_ref());
    let tagged_str = format!("{str_prefix}{str}");
    let tagged_str_bytes = tagged_str.as_bytes();
    let tagged_str_len = tagged_str_bytes.len();
    ensure!(
        tagged_str_len < 256,
        "Constant string too long: {tagged_str_len} >= 256"
    );

    let tgt_str_bytes = [&[u8::try_from(tagged_str_len)?], tagged_str_bytes].concat();
    let tgt_str_value = ctx.iw_context().const_string(&tgt_str_bytes, false);

    let global = {
        // we need a unique name for the global. We use a prefix(of length
        // 10) of the string, followed by the hash of the whole string.
        //
        // This should be enough, but for absolute guarantee of success we
        // append ".i", i an integer and find an available name.
        let hash = {
            let mut s = DefaultHasher::default();
            tgt_str_bytes.hash(&mut s);
            let h = s.finish();
            let lower32 = h as u32;
            format!("{lower32:08X}")
        };
        // Use unicode-aware character slicing to prevent slicing through multi-byte characters
        let prefix_chars: String = str.chars().take(10).collect();
        let name_prefix = format!("{symbol_prefix}{prefix_chars}.{hash}");

        (0..)
            .find_map(|i| {
                let name = format!("{name_prefix}.{i}");
                if let Some(global) = ctx.get_current_module().get_global(&name) {
                    if let Some(v) = global.get_initializer() {
                        if v == tgt_str_value {
                            // We've found a global with the same value, use that.
                            return Some(global);
                        }
                    }
                } else {
                    let global =
                        ctx.get_current_module()
                            .add_global(tgt_str_value.get_type(), None, &name);
                    global.set_initializer(&tgt_str_value);
                    global.set_linkage(Linkage::Private);
                    global.set_constant(true);
                    return Some(global);
                }
                None
            })
            .unwrap()
    };
    Ok(ctx
        .builder()
        .build_pointer_cast(
            global.as_pointer_value(),
            ctx.iw_context().i8_type().ptr_type(AddressSpace::default()),
            "",
        )?
        .as_basic_value_enum())
}

#[cfg(test)]
mod test {
    use hugr::builder::{Dataflow, DataflowHugr};
    use hugr::extension::prelude::{self, qb_t, ConstError, EXIT_OP_ID, PANIC_OP_ID};
    use hugr::extension::PRELUDE;
    use hugr::llvm::check_emission;
    use hugr::llvm::emit::test::SimpleHugrConfig;

    use hugr::llvm::test::llvm_ctx;
    use hugr::types::TypeArg;
    use rstest::rstest;

    use super::*;

    #[rstest]
    fn test_panic_emit() {
        let mut llvm_ctx = llvm_ctx(0);
        let prelude_codegen = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| ceb.add_prelude_extensions(prelude_codegen.clone()));

        // Create a hugr that has a panic message
        let error_val = ConstError::new(42, "PANIC");
        let type_arg_q = TypeArg::Runtime(qb_t());
        let type_arg_2q = TypeArg::List(vec![type_arg_q.clone(), type_arg_q]);
        let panic_op = PRELUDE
            .instantiate_extension_op(&PANIC_OP_ID, [type_arg_2q.clone(), type_arg_2q.clone()])
            .unwrap();

        let hugr = SimpleHugrConfig::new()
            .with_ins(vec![qb_t(), qb_t()])
            .with_outs(vec![qb_t(), qb_t()])
            .with_extensions(prelude::PRELUDE_REGISTRY.to_owned())
            .finish(|mut builder| {
                let [q0, q1] = builder.input_wires_arr();
                let err = builder.add_load_value(error_val);
                let [q0, q1] = builder
                    .add_dataflow_op(panic_op, [err, q0, q1])
                    .unwrap()
                    .outputs_arr();
                builder.finish_hugr_with_outputs([q0, q1]).unwrap()
            });

        check_emission!(hugr, llvm_ctx);
    }

    #[rstest]
    fn test_exit_emit() {
        let mut llvm_ctx = llvm_ctx(0);
        let prelude_codegen = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| ceb.add_prelude_extensions(prelude_codegen.clone()));

        let error_val = ConstError::new(42, "EXIT");
        let type_arg_q: TypeArg = TypeArg::Runtime(qb_t());
        let type_arg_2q: TypeArg = TypeArg::List(vec![type_arg_q.clone(), type_arg_q]);
        let exit_op = PRELUDE
            .instantiate_extension_op(&EXIT_OP_ID, [type_arg_2q.clone(), type_arg_2q.clone()])
            .unwrap();

        let hugr = SimpleHugrConfig::new()
            .with_ins(vec![qb_t(), qb_t()])
            .with_outs(vec![qb_t(), qb_t()])
            .with_extensions(prelude::PRELUDE_REGISTRY.to_owned())
            .finish(|mut builder| {
                let [q0, q1] = builder.input_wires_arr();
                let err = builder.add_load_value(error_val);
                let [q0, q1] = builder
                    .add_dataflow_op(exit_op, [err, q0, q1])
                    .unwrap()
                    .outputs_arr();
                builder.finish_hugr_with_outputs([q0, q1]).unwrap()
            });

        check_emission!(hugr, llvm_ctx);
    }
}
