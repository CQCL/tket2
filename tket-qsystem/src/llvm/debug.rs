//! LLVM lowering implementations for "tket.debug" extension.

use crate::llvm::prelude::emit_global_string;
use anyhow::{anyhow, bail, Result};
use hugr::extension::simple_op::MakeExtensionOp;
use hugr::llvm::custom::CodegenExtension;
use hugr::llvm::emit::{EmitFuncContext, EmitOpArgs};
use hugr::llvm::inkwell::AddressSpace;
use hugr::llvm::CodegenExtsBuilder;
use hugr::ops::ExtensionOp;
use hugr::{HugrView, Node};
use tket::extension::debug::{StateResult, DEBUG_EXTENSION_ID, STATE_RESULT_OP_ID};

use super::array_utils::{struct_1d_arr_alloc, struct_1d_arr_ptr_t, ArrayLowering, ElemType};

static TAG_PREFIX: &str = "USER:";

/// Codegen extension for debug functionality.
#[derive(Default)]
pub struct DebugCodegenExtension<AL: ArrayLowering> {
    array_lowering: AL,
}

impl<AL: ArrayLowering> DebugCodegenExtension<AL> {
    /// Creates a new [DebugCodegenExtension] with specified array lowering.
    pub const fn new(array_lowering: AL) -> Self {
        Self { array_lowering }
    }
}

impl<AL: ArrayLowering> CodegenExtension for DebugCodegenExtension<AL> {
    fn add_extension<'a, H: HugrView<Node = Node> + 'a>(
        self,
        builder: CodegenExtsBuilder<'a, H>,
    ) -> CodegenExtsBuilder<'a, H>
    where
        Self: 'a,
    {
        builder.extension_op(DEBUG_EXTENSION_ID, STATE_RESULT_OP_ID, {
            move |context, args| self.emit_state_result(context, args)
        })
    }
}

impl<AL: ArrayLowering> DebugCodegenExtension<AL> {
    /// Lower the `debug` `StateResult` op.
    fn emit_state_result<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
    ) -> Result<()> {
        let builder = ctx.builder();

        // Types (qubits are just i64s).
        let iw_ctx = ctx.iw_context();
        let void_t = iw_ctx.void_type();
        let i8_ptr_t = iw_ctx.i8_type().ptr_type(AddressSpace::default());
        let i64_t = iw_ctx.i64_type();

        // Tag arguments.
        let state_result = StateResult::from_extension_op(args.node().as_ref())?;
        let tag = state_result.tag;
        if tag.is_empty() {
            bail!("Empty state result tag received");
        }
        let tag_ptr = emit_global_string(ctx, tag, "res_", format!("{TAG_PREFIX}STATE:"))?;
        let tag_len = {
            let mut l = builder
                .build_load(tag_ptr.into_pointer_value(), "tag_len")?
                .into_int_value();
            if l.get_type() != i64_t {
                l = builder.build_int_z_extend(l, i64_t, "tag_len")?;
            }
            l
        };

        // Qubit array argument.
        let array_len = state_result.num_qubits;
        let [qubits] = args
            .inputs
            .try_into()
            .map_err(|_| anyhow!(format!("StateResult expects a qubit array argument")))?;
        let qubits_array = self.array_lowering.array_to_ptr(builder, qubits)?;
        let (qubits_ptr, _) = struct_1d_arr_alloc(
            iw_ctx,
            builder,
            array_len.try_into()?,
            &ElemType::Int,
            qubits_array,
        )?;

        // Build the function call.
        let fn_state_result = ctx.get_extern_func(
            "print_state_result",
            void_t.fn_type(
                &[
                    i8_ptr_t.into(),
                    i64_t.into(),
                    struct_1d_arr_ptr_t(iw_ctx, &ElemType::Int).into(),
                ],
                false,
            ),
        )?;

        builder.build_call(
            fn_state_result,
            &[tag_ptr.into(), tag_len.into(), qubits_ptr.into()],
            "print_state_result",
        )?;
        args.outputs.finish(builder, [qubits])
    }
}

#[cfg(test)]
#[allow(deprecated)] // TODO remove after switching to new array lowering
mod test {
    use tket::extension::debug::StateResult;

    use hugr::extension::simple_op::MakeRegisteredOp;
    use hugr::llvm::check_emission;
    use hugr::llvm::test::llvm_ctx;
    use hugr::llvm::test::single_op_hugr;
    use hugr::llvm::test::TestContext;

    use crate::llvm::array_utils::DEFAULT_HEAP_ARRAY_LOWERING;
    use crate::llvm::array_utils::DEFAULT_STACK_ARRAY_LOWERING;
    use crate::llvm::prelude::QISPreludeCodegen;

    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case::state_result(1, StateResult::new("test_state_result".to_string(), 2), &DEFAULT_STACK_ARRAY_LOWERING)]
    #[case::state_result(2, StateResult::new("test_state_result".to_string(), 2), &DEFAULT_HEAP_ARRAY_LOWERING)]
    fn emit_debug_codegen(
        #[case] _i: i32,
        #[with(_i)] mut llvm_ctx: TestContext,
        #[case] op: StateResult,
        #[case] array_lowering: &'static (impl ArrayLowering + Clone),
    ) {
        let pcg = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| {
            ceb.add_extension(DebugCodegenExtension::new(array_lowering.clone()))
                .add_extension(array_lowering.codegen_extension())
                .add_prelude_extensions(pcg.clone())
                .add_default_int_extensions()
                .add_float_extensions()
        });
        let ext_op = op.to_extension_op().unwrap().into();
        let hugr = single_op_hugr(ext_op);
        check_emission!(hugr, llvm_ctx);
    }
}
