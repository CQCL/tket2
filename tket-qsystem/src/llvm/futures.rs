//! LLVM lowering implementations for "tket.futures" operations.

use crate::extension::futures::{self, FutureOp, FutureOpDef, FUTURE_TYPE_NAME};
use anyhow::{anyhow, Result};
use hugr::extension::prelude::bool_t;
use hugr::extension::simple_op::MakeExtensionOp;
use hugr::ops::{ExtensionOp, Value};
use hugr::std_extensions::arithmetic::int_types::INT_TYPES;
use hugr::types::TypeArg;
use hugr::{HugrView, Node};
use hugr_llvm::custom::CodegenExtension;
use hugr_llvm::emit::func::EmitFuncContext;
use hugr_llvm::emit::{emit_value, EmitOpArgs};
use hugr_llvm::CodegenExtsBuilder;
use inkwell::builder::Builder;
use inkwell::context::Context;
use inkwell::types::{BasicTypeEnum, IntType};
use inkwell::values::FunctionValue;
use tket::hugr::{
    self,
    llvm::{self as hugr_llvm, inkwell},
};

/// Codegen extension for futures
pub struct FuturesCodegenExtension;

impl CodegenExtension for FuturesCodegenExtension {
    fn add_extension<'a, H: HugrView<Node = Node> + 'a>(
        self,
        builder: CodegenExtsBuilder<'a, H>,
    ) -> CodegenExtsBuilder<'a, H>
    where
        Self: 'a,
    {
        builder
            .custom_type(
                (futures::EXTENSION_ID, FUTURE_TYPE_NAME.to_owned()),
                |session, hugr_type| {
                    match (hugr_type.name().as_str(), hugr_type.args()) {
                        // For now, we only support future bools
                        ("Future", [TypeArg::Runtime(ty)]) if *ty == bool_t() => {
                            Ok(future_type(session.iw_context()))
                        }
                        ("Future", [TypeArg::Runtime(ty)]) if *ty == INT_TYPES[6] => {
                            Ok(future_type(session.iw_context()))
                        }
                        _ => Err(anyhow!(
                            "FuturesCodegenExtension: Unsupported type: {}",
                            hugr_type
                        )),
                    }
                },
            )
            .simple_extension_op::<FutureOpDef>(|context, args, _op| {
                FuturesEmitter(context).emit(args)
            })
    }
}

/// The LLVM type representing a future handle.
pub fn future_type(context: &Context) -> BasicTypeEnum<'_> {
    // The runtime represent a future handle as an i64
    context.i64_type().into()
}

struct FuturesEmitter<'c, 'd, 'e, H: HugrView<Node = Node>>(&'d mut EmitFuncContext<'c, 'e, H>);

impl<'c, H: HugrView<Node = Node>> FuturesEmitter<'c, '_, '_, H> {
    fn iw_context(&self) -> &'c Context {
        self.0.typing_session().iw_context()
    }

    fn ll_future_type(&self) -> BasicTypeEnum<'c> {
        future_type(self.iw_context())
    }

    fn ll_bool_type(&self) -> IntType<'c> {
        self.iw_context().bool_type()
    }

    fn ll_uint_type(&self) -> IntType<'c> {
        self.iw_context().i64_type()
    }

    fn builder(&self) -> &Builder<'c> {
        self.0.builder()
    }

    fn get_func_inc_refcount(&self) -> Result<FunctionValue<'c>> {
        let func_type = self
            .iw_context()
            .void_type()
            .fn_type(&[self.ll_future_type().into()], false);
        self.0.get_extern_func("___inc_future_refcount", func_type)
    }

    fn get_func_dec_refcount(&self) -> Result<FunctionValue<'c>> {
        let func_type = self
            .iw_context()
            .void_type()
            .fn_type(&[self.ll_future_type().into()], false);
        self.0.get_extern_func("___dec_future_refcount", func_type)
    }

    fn get_func_read_bool(&self) -> Result<FunctionValue<'c>> {
        let func_type = self
            .ll_bool_type()
            .fn_type(&[self.ll_future_type().into()], false);
        self.0.get_extern_func("___read_future_bool", func_type)
    }

    fn get_func_read_uint(&self) -> Result<FunctionValue<'c>> {
        let func_type = self
            .ll_uint_type()
            .fn_type(&[self.ll_future_type().into()], false);
        self.0.get_extern_func("___read_future_uint", func_type)
    }

    fn emit(&mut self, args: EmitOpArgs<'c, '_, ExtensionOp, H>) -> Result<()> {
        let future_op = FutureOp::from_optype(&args.node().generalise()).unwrap();
        let op = &future_op.op;
        let typ = &future_op.typ;
        match (op, typ) {
            (FutureOpDef::Dup, _) => {
                let [arg] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("Dup expects a single input"))?;
                let func = self.get_func_inc_refcount()?;
                self.builder()
                    .build_call(func, &[arg.into()], "inc_refcount")?;
                args.outputs.finish(self.builder(), [arg, arg])
            }
            (FutureOpDef::Free, _) => {
                let [arg] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("Free expects a single input"))?;
                let func = self.get_func_dec_refcount()?;
                self.builder()
                    .build_call(func, &[arg.into()], "dec_refcount")?;
                args.outputs.finish(self.builder(), [])
            }
            (FutureOpDef::Read, ty) if *ty == bool_t() => {
                let [arg] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("Read expects a single input"))?;
                let read_func = self.get_func_read_bool()?;
                let result_i1 = self
                    .builder()
                    .build_call(read_func, &[arg.into()], "read_bool")?
                    .try_as_basic_value()
                    .unwrap_left()
                    .into_int_value();
                let dec_refcount_func = self.get_func_dec_refcount()?;
                self.builder()
                    .build_call(dec_refcount_func, &[arg.into()], "dec_refcount")?;
                let true_val = emit_value(self.0, &Value::true_val())?;
                let false_val = emit_value(self.0, &Value::false_val())?;
                let result = self
                    .builder()
                    .build_select(result_i1, true_val, false_val, "measure")?;
                args.outputs.finish(self.builder(), [result])
            }
            (FutureOpDef::Read, ty) if *ty == INT_TYPES[6] => {
                let [arg] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("Read expects a single input"))?;
                let read_func = self.get_func_read_uint()?;
                let result = self
                    .builder()
                    .build_call(read_func, &[arg.into()], "read_uint")?
                    .try_as_basic_value()
                    .unwrap_left();
                let dec_refcount_func = self.get_func_dec_refcount()?;
                self.builder()
                    .build_call(dec_refcount_func, &[arg.into()], "dec_refcount")?;
                args.outputs.finish(self.builder(), [result])
            }
            _ => Err(anyhow!(
                "Unsupported future operation: {:?} with type: {}",
                op,
                typ
            )),
        }
    }
}

#[cfg(test)]
mod test {
    use hugr::extension::simple_op::MakeRegisteredOp;
    use hugr::std_extensions::arithmetic::int_types::int_type;
    use hugr_llvm::check_emission;
    use hugr_llvm::test::llvm_ctx;
    use hugr_llvm::test::single_op_hugr;
    use hugr_llvm::test::TestContext;

    use super::*;
    #[rstest::rstest]
    #[case::read_bool(1,FutureOp { op: FutureOpDef::Read, typ: bool_t() })]
    #[case::dup_bool(2,FutureOp { op: FutureOpDef::Dup, typ: bool_t() })]
    #[case::free_bool(3,FutureOp { op: FutureOpDef::Free, typ: bool_t() })]
    #[case::read_int(4,FutureOp { op: FutureOpDef::Read, typ: int_type(6) })]
    #[case::dup_int(5,FutureOp { op: FutureOpDef::Dup, typ: int_type(6) })]
    #[case::free_int(6,FutureOp { op: FutureOpDef::Free, typ: int_type(6) })]
    fn emit_futures_codegen(
        #[case] _i: i32,
        #[with(_i)] mut llvm_ctx: TestContext,
        #[case] op: FutureOp,
    ) {
        llvm_ctx.add_extensions(|ceb| {
            ceb.add_extension(FuturesCodegenExtension)
                .add_default_int_extensions()
        });
        let hugr = single_op_hugr(op.to_extension_op().unwrap().into());
        check_emission!(hugr, llvm_ctx);
    }
}
