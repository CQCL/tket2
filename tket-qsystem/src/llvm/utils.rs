//! LLVM codegen for the `tket.qsystem.utils` extension.
use tket::hugr::{self};

use crate::extension::utils::UtilsOp;
use anyhow::Result;
use hugr::llvm::custom::CodegenExtension;
use hugr::llvm::emit::func::EmitFuncContext;
use hugr::llvm::emit::EmitOpArgs;
use tket::hugr::ops::ExtensionOp;
use tket::hugr::{HugrView, Node};

/// Codegen extension for `tket_qsystem` `utils` extension.
pub struct UtilsCodegenExtension;

impl CodegenExtension for UtilsCodegenExtension {
    fn add_extension<'a, H: HugrView<Node = Node> + 'a>(
        self,
        builder: tket::hugr::llvm::CodegenExtsBuilder<'a, H>,
    ) -> tket::hugr::llvm::CodegenExtsBuilder<'a, H>
    where
        Self: 'a,
    {
        builder.simple_extension_op(|context, args, op| emit_utils_op(context, args, op))
    }
}

/// Lower the `tket_qsystem` `utils` extension.
fn emit_utils_op<H: HugrView<Node = Node>>(
    ctx: &EmitFuncContext<'_, '_, H>,
    args: EmitOpArgs<'_, '_, ExtensionOp, H>,
    op: UtilsOp,
) -> Result<()> {
    match op {
        UtilsOp::GetCurrentShot => {
            let fn_get_cur_shot = ctx.get_extern_func(
                // fun get_current_shot() -> uint
                "get_current_shot",
                ctx.typing_session()
                    .iw_context()
                    .i64_type()
                    .fn_type(&[], false),
            )?;
            let result = ctx
                .builder()
                .build_call(fn_get_cur_shot, &[], "shot")?
                .try_as_basic_value()
                .unwrap_left();
            args.outputs.finish(ctx.builder(), [result])
        }
    }
}

#[cfg(test)]
mod test {
    use crate::extension::utils::UtilsOp;
    use hugr::extension::simple_op::MakeRegisteredOp;
    use hugr::llvm::check_emission;
    use hugr::llvm::test::llvm_ctx;
    use hugr::llvm::test::single_op_hugr;
    use hugr::llvm::test::TestContext;
    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case::get_current_shot(1, UtilsOp::GetCurrentShot)]
    fn emit_utils_codegen(
        #[case] _i: i32,
        #[with(_i)] mut llvm_ctx: TestContext,
        #[case] op: UtilsOp,
    ) {
        llvm_ctx.add_extensions(|ceb| {
            ceb.add_extension(UtilsCodegenExtension)
                .add_default_int_extensions()
        });
        let ext_op = op.to_extension_op().unwrap().into();
        let hugr = single_op_hugr(ext_op);
        check_emission!(hugr, llvm_ctx);
    }
}
