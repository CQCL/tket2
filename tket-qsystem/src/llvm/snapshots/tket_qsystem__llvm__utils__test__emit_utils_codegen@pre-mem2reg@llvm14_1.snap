---
source: tket-qsystem/src/llvm/utils.rs
expression: mod_str
---
; ModuleID = 'test_context'
source_filename = "test_context"

define private i64 @_hl.main.1() {
alloca_block:
  %"0" = alloca i64, align 8
  %"4_0" = alloca i64, align 8
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  %shot = call i64 @get_current_shot()
  store i64 %shot, i64* %"4_0", align 4
  %"4_01" = load i64, i64* %"4_0", align 4
  store i64 %"4_01", i64* %"0", align 4
  %"02" = load i64, i64* %"0", align 4
  ret i64 %"02"
}

declare i64 @get_current_shot()
